src/economy/shops/make-shop.ts

```diff
@@ -0,0 +1,467 @@
+import { color, text } from 'craftjs-plugin/chat';
+import { TextComponent, TranslatableComponent } from 'net.md_5.bungee.api.chat';
+import { Bukkit, ChatColor, OfflinePlayer } from 'org.bukkit';
+import { Block, Sign } from 'org.bukkit.block';
+import { Chest, WallSign } from 'org.bukkit.block.data.type';
+import { Player } from 'org.bukkit.entity';
+import { Action, SignChangeEvent } from 'org.bukkit.event.block';
+import { PlayerInteractEvent } from 'org.bukkit.event.player';
+import { EquipmentSlot, ItemStack } from 'org.bukkit.inventory';
+import { ChatMessage, GLOBAL_PIPELINE } from '../../chat/pipeline';
+import { dataView } from '../../common/datas/view';
+import { getItemName, getCustomTranslation } from '../../common/helpers/items';
+import { coinModelIdToCurrencyId, getCoinData } from '../money-mold';
+import { ShopData } from './ShopData';
+import { Currency } from '../currency';
+import { distanceBetween } from '../../common/helpers/locations';
+import { errorMessage } from '../../chat/system';
+import { getTaxCollector, getTaxes } from './taxes';
+import { getBlockBehind } from './helpers';
+export type ShopType = 'SELLING' | 'BUYING';
+
+interface ShopInfo {
+  type?: ShopType;
+  item?: ItemStack;
+  price?: number;
+  currency?: Currency;
+  taxRate?: number;
+  taxCollector?: OfflinePlayer;
+}
+
+export interface ValidShopInfo {
+  type: ShopType;
+  item: ItemStack;
+  price: number;
+  currency: Currency;
+  taxRate: number;
+  taxCollector?: OfflinePlayer;
+}
+
+/**
+ * Steps during the shop making session
+ */
+type Step =
+  | 'START'
+  | 'SET_ITEM'
+  | 'SET_PRICE'
+  | 'SET_CURRENCY'
+  | 'SET_TAXES'
+  | 'SET_TAX_COLLECTOR';
+
+interface ShopMakingSession {
+  sign: Block;
+  step: Step;
+  shopInfo: ShopInfo;
+  updated: Date;
+}
+
+/**
+ * A Map containing players who are making a shop currently, and information about the session
+ */
+const sessions = new Map<Player, ShopMakingSession>();
+
+/**
+ * Stop shop making sessions if the player has been idle or moved far away from the shop
+ */
+const SHOP_MAKING_TIMEOUT_MS = 1000 * 10;
```

Comment:
Maybe a bit longer timeout, 10 seconds is quite short for someone creating a shop for first time.